indexbuilder.go

```diff
@@ -559,3 +514,61 @@ func (b *IndexBuilder) branchMask(br string) uint64 {
 	}
 	return 0
 }
+
+type DocChecker struct {
+	// A map to count the unique trigrams in a doc. Reused across docs to cut down on allocations.
+	trigrams map[ngram]struct{}
+}
+
+// Check returns a reason why the given contents are probably not source texts.
+func (t *DocChecker) Check(content []byte, maxTrigramCount int, allowLargeFile bool) error {
+	if len(content) == 0 {
+		return nil
+	}
+
+	if len(content) < ngramSize {
+		return fmt.Errorf("file size smaller than %d", ngramSize)
+	}
+
+	if index := bytes.IndexByte(content, 0); index > 0 {
+		return fmt.Errorf("binary data at byte offset %d", index)
+	}
+
+	// PERF: we only need to do the trigram check if the upperbound on content is greater than
+	// our threshold. Also skip the trigram check if the file is explicitly marked as allowed.
+	if trigramsUpperBound := len(content) - ngramSize + 1; trigramsUpperBound <= maxTrigramCount || allowLargeFile {
+		return nil
+	}
+
+	var cur [3]rune
+	byteCount := 0
+	t.clearTrigrams(maxTrigramCount)
+
+	for len(content) > 0 {
```

Comment:
This seemed like a natural fit for a cardinality sketch like hyperloglog. I tried that out using the `axiomhq/hyperloglog` library, and it turned out to be way slower than what we currently do. So it's not worth it given the relatively low set size (bounded by maxTrigramCount) and the large number of times we need to check this.

Reply:
I thought of the same DS. Surprising it isn't faster, could it just be the implementation?

Reply:
Perhaps! That looked like the most widely used implementation in the Go world ... maybe something to revisit though.